field/TinyLang/Field/Raw/Parser.hs

```diff
@@ -0,0 +1,339 @@
+analysis and grammar we will explicitly mark all lexical tokens.
+
+Please note that expressions are the designated start production for
+the parser.
+
+We are using the following convention
+
+* /tokens/ are enclosed in parentheses @""@.
+* /regular expressions/ are enclosed in square brackets @[]@.
+
+== Literals
+
+Literals are defined as follows
+
+@
+bool-literal ::= "T" | "F"
+vec-literal  ::= "{" (bool-literal ("," bool-literal)*)? "}"
+int-literal  ::= [0-9]+
+@
+
+== Keywords
+
+Here is a list of keywords used in the language.
+
+@
+keyword ::=
+    "and"
+    "assert"
+    "do"
+    "else"
+    "end"
+    "for"
+    "if"
+    "inv"
+    "let"
+    "neq0"
+    "not"
+    "or"
+    "then"
+    "to"
+    "unpack"
+    "xor"
+@
+
+== Identifiers
+
+To avoid a syntactic clash with /bool-literals/, identifiers for field
+variables start with a lower-case letter.  Identifiers for boolean
+variables are prefiex by @?@ and identifiers for vector variables are
+prefixed by @#@.  We will check if an identifier is not a keyword.
+
+@
+ident ::= ( "?" | "#" ) [a-z] ([a-z0-9_'])*
+@
+
+== Constants
+
+At the moment constants consist of literals only.
+
+@
+const ::=
+    bool-literal
+    int-literal
+    vec-literal
+@
+
+== Expressions
+
+As mentioned above @expr@ is the designated start production for the
+parser.
+
+@
+expr ::=
+    const
+    "(" expr ")"
+    ident
+    expr infix-op expr
+    prefix-op expr
+    expr "[" expr "]"
+    statement ";" expr
+    "if" expr "then" expr "else" expr
+
+infix-op ::=
+    "and"
+    "or"
+    "xor"
+    "=="
+    "<="
+    "<"
+    ">="
+    ">"
+    "+"
+    "-"
+    "*"
+    "/"
+
+prefix-op ::=
+    "not"
+    "neq0"
+    "neg"
+    "inv"
+    "unpack"
+@
+
+== Statement
+
+Statements are a bit odd at the moment, as they are neither
+expressions nor the usual statements.
+
+@
+statement ::=
+    "let" var "=" expr
+    "assert" expr
+    "for" var "=" int "to" int "do" statements "end"
+
+statements ::=
+    (statement (";" statement)*)?
+@
+
+== Operator Precedence
+Precedence: "not" > "xor" > "and" > "or" (but use parentheses anyway).
+if-then-else has to be parenthesised unless it's at the very top.
+
+Precedence for numeric operators is standard:  {neg,inv} > {*,/} > {+,- }.
+Things like "neg inv 5" are illegal: use parentheses.
+
+The code is based on the tutorial at
+https://markkarpov.com/megaparsec/parsing-simple-imperative-language.html
+
+See also https://markkarpov.com/megaparsec/megaparsec.html
+-}
+
+
+module TinyLang.Field.Raw.Parser
+    ( parseRational
+    , p
+    ) where
+
+import           TinyLang.Prelude hiding ( option
+                                         , many
+                                         , try
+                                         )
+
+import           TinyLang.Field.Raw.Core
+import           TinyLang.Field.Existential
+import           TinyLang.Field.UniVal
+import           Data.Set ( fromList
```

Comment:
We have a convention that imports from external libraries and imports of internal modules are separated by a newline. The latter go first and the former go next (after a newline). This makes it easier to read imports.